### PR TITLE
⚠️ Export scaffolding machinery

### PR DESCRIPTION
**Pull request by Adirio**
It is technically a breaking change (⚠️) because some errors that were previously placed in `pkg/model` and `pkg/model/file` have been moved to `pkg/machinery` and no longer expose `NewXxxxxxError` and `IsXxxxxxxError` functions, they expose themselves to use `errors.As` and `errors.Is`. It also removes the `NewUniverse` method and the `WithXxxxx` options are now part of the `NewScaffold` method, instead of having to pass a universe to every `Execute` call.

Closes: #2036
Blocked by: #2080

### CODE REVIEW

**Review by camilamacedo86 — APPROVED** [review]
The change which was significant and was discussed in the meeting is: https://github.com/kubernetes-sigs/kubebuilder/pull/2080. (which was approved and merged by Eric, who raised the concerns. )

This one is the only flag as breaking change because is moving things around and changing the names. However, the same feature options still in place. Also, it shows very small changes which means that shows fine we move forward here.

PS.: We also agree in the meetings/issues to export the machinery to remove this duplication from the consumer (SDK)

/lgtm

File: pkg/machinery/scaffold.go [hunk 1]
```diff
@@ -41,32 +49,96 @@ var options = imports.Options{
 }
 
 // Scaffold uses templates to scaffold new files
-type Scaffold interface {
```
Comment (estroz) [inline pkg/machinery/scaffold.go:44]: Why was this changed to a struct?

File: pkg/machinery/scaffold.go [hunk 2]
```diff
@@ -375,16 +445,29 @@ func (s scaffold) writeFile(f *file.File) error {
...
+
+	// Create or truncate the file
+	writer, err := s.fs.OpenFile(f.Path, createOrUpdate, s.filePerm)
 	if err != nil {
-		return err
+		return CreateFileError{err}
 	}
+	defer func() {
+		if closeErr := writer.Close(); err == nil && closeErr != nil {
+			err = CloseFileError{err}
```
Comment (estroz) [inline pkg/machinery/scaffold.go:464]: This overwrites `err` in case `Write()` fails. Should be a `log.Error(closeErr)` instead.

File: pkg/machinery/scaffold.go (hunk 1 above)
Comment (Adirio) [reply pkg/machinery/scaffold.go:44]: Because we were exporting a `Option` that was a function on an unexported struct.

File: pkg/machinery/scaffold.go (hunk 2 above)
Comment (Adirio) [reply pkg/machinery/scaffold.go:464]: We check that `err == nil`, so we only overwrite in case there was no error.

Comment (estroz) [reply pkg/machinery/scaffold.go:464]: Right, missed that bit. What if `err != nil && closeErr != nil`?

File: pkg/machinery/scaffold.go (hunk 1 above)
Comment (estroz) [reply pkg/machinery/scaffold.go:44]: You can do `type ScaffoldOption func(*scaffold)` since the predefined options have access to the internal type.

Comment (estroz) [reply pkg/machinery/scaffold.go:44]: I would prefer ^ but it's an unimportant design decision.

File: pkg/machinery/scaffold.go (hunk 2 above)
Comment (Adirio) [reply pkg/machinery/scaffold.go:464]: `err` is reported as it happened before. Same as it was before.

File: pkg/machinery/scaffold.go (hunk 1 above)
Comment (Adirio) [reply pkg/machinery/scaffold.go:44]: That means that we are sporting `ScaffoldOption` which has a unexported parameter. I avoided this when able.

Comment (estroz) [reply pkg/machinery/scaffold.go:44]: That's a common practice. The set of options is fixed because `Scaffold`'s fields are unexported anyway.

Comment (Adirio) [reply pkg/machinery/scaffold.go:44]: Oh, yes, I know. The result is exactly the same. But there was not added value by the interface either. And this way we don't have a unexported type in pkg.go.dev.
